### PR TITLE
feat: add OperationalError cause property

### PR DESCRIPTION
**Pull request by andygout**
FTDCS-250

Jira card: https://financialtimes.atlassian.net/jira/software/c/projects/FTDCS/boards/1426?modal=detail&selectedIssue=FTDCS-250

> **Rowan Manning:** @andygout [pointed out](https://financialtimes.slack.com/archives/CPNKK12S1/p1659013086713049?thread_ts=1659003590.492259&cid=CPNKK12S1) in one of my examples for handling errors we lose some information about the original error. I’ve had a potential solution in mind for this for a while and it’s worth getting down on paper.

> I think that OperationalError should have a new first-class property named cause. This expects another Error object and marks that error as the root cause of the operational error. This will allow us to use generic messages in some places but still capture the original error information. Using the example from the error handling docs, we could update it to be:

```js
try {
    fruit.stockLevel = await fruitApiClient.getStockLevel(fruit.id);
} catch (error) {
    console.warn({
        event: 'RECOVERABLE_ERROR',
        error: serializeError(new OperationalError({
            code: 'FRUIT_STOCK_LEVEL_FAILED',
            message: `The Fruit API did not return a stock level for ID ${fruit.id}`,
            relatesToSystems: ['fruit-api'],
            cause: error // This is the new code
        }))
    });
    fruit.stockLevel = null;
}
```

> We’d also want to update our logger so that it exposes this information and also serialises any errors on the cause property, e.g. the log output could change to:

```js
{
    event: 'RECOVERABLE_ERROR',
    message: 'OperationalError: The Fruit API did not return a stock level for ID X',

    error: {
        // missing some properties for brevity
        name: 'OperationalError',
        code: 'FRUIT_STOCK_LEVEL_FAILED',
        message: 'The Fruit API did not return a stock level for ID X',
        
        cause: {
            name: 'Error',
            code: null,
            message: 'HTTP 503'
        }
    },

    app: {
        // etc
    }
}
```

### CODE REVIEW

File: docs/getting-started/handling-errors.md [hunk 1]
```diff
@@ -337,10 +337,10 @@ app.get('/fruit/:name', async (request, response, next) => {
             console.warn({
                 event: 'RECOVERABLE_ERROR',
                 error: serializeError(new OperationalError({
-                    statusCode: error.statusCode,
                     code: 'FRUIT_STOCK_LEVEL_FAILED',
                     message: `The Fruit API did not return a stock level for ID ${fruit.id}`,
-                    relatesToSystems: ['fruit-api']
+                    relatesToSystems: ['fruit-api'],
+                    cause: error
```
Comment (rowanmanning) [inline docs/getting-started/handling-errors.md:343]: **praise:** nice 👏 this is a much better example

File: docs/getting-started/throwing-errors.md [hunk 2]
```diff
@@ -228,6 +228,20 @@ throw new OperationalError({
 });
 ```
 
+You might create an `OperationalError` instance in reaction to an error that has been caught. The root cause error and all the diagnostic information it contains can be included in the `OperationalError` instance by setting it as the value of the `cause` property:
```
Comment (rowanmanning) [inline docs/getting-started/throwing-errors.md:231]: **praise:** great explanation and example

File: packages/serialize-error/test/lib/index.spec.js [hunk 3]
```diff
@@ -84,6 +85,30 @@ describe('@dotcom-reliability-kit/serialize-error', () => {
...
+				const rootCauseErrorInstance = new Error(
+					'mock root cause error message'
+				);
+				error.cause = rootCauseErrorInstance;
+				expect(serializeError(error)).toMatchObject({
+					cause: serializeError(rootCauseErrorInstance)
+				});
+				expect(serializeError(error).cause).toMatchObject({
+					message: 'mock root cause error message'
+				});
```
Comment (andygout) [inline packages/serialize-error/test/lib/index.spec.js:99]: I could have gone to town on testing the serialisation of the `cause` error to provide the same coverage that is given for the subject (i.e. containing) error, e.g. when the `cause` value is the result of `serialiseError()` being called with:
- an error object
- en error-like object
- a string
- a number
- an array

However, the `serialiseError()` is already being tested so thoroughly by the subject error, so ensuring that the root cause error `message` value is as expected felt to be a sufficient measure for testing this.

**Comment by andygout** [comment]
> We’d also want to update our logger so that it exposes this information

From what I can tell, the `cause` value of error should be surfaced via the inclusion of the `error` property (in which it is situated) and that no changes are required to `packages/log-error/lib/index.js` but please let me know if I have misunderstood something.

File: packages/serialize-error/lib/index.js [hunk 4]
```diff
@@ -61,6 +63,11 @@ function serializeError(error) {
 		errorProperties.relatesToSystems = error.relatesToSystems;
 	}
 
+	// If set, error cause (which in turn is an error instance) is serialized
+	if (error.cause && error.cause instanceof Error) {
+		errorProperties.cause = serializeError(error.cause);
```
Comment (rowanmanning) [inline packages/serialize-error/lib/index.js:68]: **thought:** maybe down the line we need to think about whether to limit the recursion here? I don't think in this PR but I guess it's a possible vector for errors to creep in. Though tbh I think you'd have to try quite hard to make this infinitely recurse 🤷 anyway not something to address now

**Review by rowanmanning — CHANGES_REQUESTED** [review]
This is looking great, really excellent 🙂 and I think it'll make `OperationalError` a lot more useful when it starts appearing in apps.

My one change request is that we need to update the READMEs for a few of the packages:

* In the serialize-error README [here](https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/serialize-error#serializederror-type) and [here](https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/serialize-error#serializeerror) (also don't forget to update the table of contents which is manual for now)
* In the errors README [here](https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/errors#operationalerrorrelatestosystems) (also the ToC)

**Review by rowanmanning — APPROVED** [review]
Looks really great 👍 thanks Andy! Feel free to merge

**Pull request merged**